[WebAssembly] Autovec support for dot

sparker-arm opened the pull request:
Enable the use of partial.reduce.add that we can lower to dot or a tree of (add (extmul_low_u, extmul_high_u)) for the unsigned case. We support both v8i16 and v16i8 inputs.

sparker-arm commented:
Ping?

dschuff commented:
Thanks for the PR, the code looks pretty reasonable. I'm curious about the partial.reduce.add intrinsic since I'm not really familiar with it. Is this something that the autovectorizer will generate by default now (or maybe now that you've implemented `getPartialReductionCost`?)

sparker-arm commented:
> or maybe now that you've implemented getPartialReductionCost?

Correct, but with the caveat that it won't trigger at the moment due to the vectorization factor selected. For our dot, we need a factor of 8 or 16 but, AFAICT, the vectorizer will not go above 4 as we are using an i32 for the arithmetic. The tests that shows dot being generated are doing so because I've used the `-vectorizer-maximize-bandwidth` option. This hook can be implemented in the backend too, but I'm not convinced that's the best approach. I think some extra work will be needed in the vectorizer to explore a wider vector factor, based upon the input widths, for these partial reductions.

sparker-arm commented:
And thanks for taking a look!

sparker-arm committed:
[WebAssembly] Autovec support for dot

Enable the use of partial.reduce.add that we can lower to dot or a
tree of (add (extmul_low_u, extmul_high_u)) for the unsigned
case. We support both v8i16 and v16i8 inputs.

sparker-arm commented:
Ping?
